gobblin-core-base/src/main/java/gobblin/writer/AsyncWriterManager.java

```diff
@@ -0,0 +1,519 @@
+          }
+        }
+      }
+    });
+  }
+
+  private class RetryRunner implements Runnable {
+
+    private final LinkedBlockingQueue<Attempt> _retryQueue;
+
+    public RetryRunner() {
+      Preconditions.checkArgument(retryQueue.isPresent(), "RetryQueue must be present for RetryRunner");
+      _retryQueue = retryQueue.get();
+    }
+
+    @Override
+    public void run() {
+      while (true)
+      {
+        try {
+          Attempt attempt = _retryQueue.take();
+          if (attempt != null) {
+            attemptWrite(attempt);
+          }
+        } catch (InterruptedException e) {
+          log.info("Retry thread interrupted... will exit");
+          Throwables.propagate(e);
+        }
+      }
+    }
+  }
+
+  @Override
+  public void cleanup()
+      throws IOException {
+    // legacy api ...
+  }
+
+  @Override
+  public long recordsWritten() {
+    return recordsSuccess.getCount();
+  }
+
+  @Override
+  public long bytesWritten()
+      throws IOException {
+    return bytesWritten.getCount();
+  }
+
+  @Override
+  public void close()
+      throws IOException {
+    log.debug("Close called");
+    this.closer.close();
+  }
+
+  @Override
+  public void commit()
+      throws IOException {
+    log.debug("Commit called, will wait for commitTimeout : " + commitTimeoutInNanos / MILLIS_TO_NANOS + "ms");
+    long commitStartTime = System.nanoTime();
+    asyncDataWriter.flush();
+    while (((System.nanoTime() - commitStartTime) < commitTimeoutInNanos)  &&
+        (recordsIn.getCount() != (recordsSuccess.getCount() + recordsFailed.getCount())))
+    {
+      log.debug("Commit waiting... records produced: " + recordsIn.getCount() + " written: "
+          + recordsSuccess.getCount() + " failed: " + recordsFailed.getCount());
+      try {
+        Thread.sleep(commitStepWaitTimeMillis);
+      } catch (InterruptedException e) {
+        throw new IOException("Interrupted while waiting for commit to complete", e);
+      }
+    }
+    log.debug("Commit done waiting");
+    long recordsProducedFinal = recordsIn.getCount();
+    long recordsWrittenFinal = recordsSuccess.getCount();
+    long recordsFailedFinal = recordsFailed.getCount();
+    long unacknowledgedWrites  = recordsProducedFinal - recordsWrittenFinal - recordsFailedFinal;
+    long totalFailures = unacknowledgedWrites + recordsFailedFinal;
+    if (unacknowledgedWrites > 0) // timeout
+    {
+      log.warn("Timeout waiting for all writes to be acknowledged. Missing " + unacknowledgedWrites
+          + " responses out of " + recordsProducedFinal);
+    }
+    if (totalFailures > 0 && recordsProducedFinal > 0)
+    {
+      String message = "Commit failed to write " + totalFailures
+          + " records (" + recordsFailedFinal + " failed, " + unacknowledgedWrites + " unacknowledged) out of "
+          + recordsProducedFinal + " produced.";
+      double failureRatio = (double)totalFailures / (double)recordsProducedFinal;
```

Comment:
Ok to do this later when isFailureFatal actually evolves to being able to do more sophisticated analysis? If I just move this logic to isFailureFatal now, it will run the risk of early termination if there are a large number of failures early (e.g. if the 2nd write fails, we're already at 50% failures). Also unacknowledged writes are treated as failures after timeout on commit.. cannot do the same when writes are still in flight.